src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/HostProject.cs

```diff
@@ -1,28 +1,15 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT license. See License.txt in the project root for license information.
 
-using System;
 using System.IO;
 using Microsoft.AspNetCore.Razor.Language;
 using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.Extensions.Internal;
 
 namespace Microsoft.CodeAnalysis.Razor.ProjectSystem;
 
-internal class HostProject
+internal record class HostProject
```

Comment:
Just for interest, I'm trying to think what benefit we get from using a record when we have to override Equals and GetHashCode? Is it just the `with` operator?

Reply:
(Annoying nit: `with` is not really an operator Instead, we refer to the syntax as a "with expression".)

Hmmm... the structure of your feedback leads me to believe that you might not think that the `with` expression is worth it. I'm feeling called upon to enumerate the virtues of records, and I'm happy to oblige! 😁 

Moving a simply immutable class to a record provides several benefits:

- **`with` Expression Support**
  Yes, we get `with` expression support, which is super helpful! While working on this change, I encountered a lot of code that creates one `HostProject` from another by passing _all_ of the data into the constructor with just one change. All of that construction boilerplate is what `with` expressions were specifically designed to deal with. (i.e., that thing we use a high-flatulent term to describe: "non-destructive mutation".)
 
  Ultimately, `with` expressions can make code more maintainable. Back when you added `HostProject.DisplayName`, you took the approach of adding an optional constructor parameter so that you only had to update particular constructor calls. That was a great choice and allowed you to avoid a lot of pain. However, if it hadn't been possible to derive `DisplayName` from the file path, you would've had a load of mindless, mechanical work ahead to update all of those constructor call-sites to pass `DisplayName` along. Of course, you could have worked around that by adding `HostProject.WithXXX(...)` methods for the properties that can be changed non-destructively. That would achieve the same maintainability goals, and that's why `with` expressions are there for you.

  FWIW, this PR only updates _some_ of the obvious places where a `with` expression could be employed with `HostProject` and `HostDocument`, but there are others, mostly in tests. I didn't dig too deeply before submitting this PR, but I intend to submit a follow up.

- **Value Equality**
  While it was necessary to provide `Equals(...)` and `GetHashCode()` to ensure equality is handled correctly for file paths, the record machinery handles all of the extra boilerplate to provide value equality. I didn't have to override `bool Equals(object? o)`. I didn't have to add the implementation of `IEquatable<T>`. I didn't have to add operator overloads for `==` and `!=`. All of that is provided.

- **Inheritance Support**
  I was sad to find that `HostProject` had two subtypes, so I couldn't seal it. However, record inheritance ensures that the same runtime will be used when a `with` expression creates a new instance. So, if the left side of `with` is a `FallbackHostProject` at runtime, a new `FallbackHostProject` will be created.

- **Improved `ToString()` implementation**
  The default `ToString()` implementation for a record shows the property names and values rather than the fully-qualified type name. So, the debugging experience is immediately better without having to override `ToString()` ourselves.

There are other benefits of records that we _don't_ get. For example, declaring a nominal record rather than a positional record means that a `Deconstruct` method isn't generated for us, since the property order isn't known.

Reply:
> you might not think that the `with` expression is worth it

I would not go that far, but I can see how you could get that impression from my comment. Thank you for the explanation, I definitely was not thinking about inheritance in particular.